src/EditorFeatures/Core/Implementation/FindReferences/IDefinitionsAndReferencesFactory.cs

```diff
@@ -0,0 +1,271 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.CodeAnalysis.Completion;
+using Microsoft.CodeAnalysis.FindSymbols;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.FindReferences
+{
+    internal interface IDefinitionsAndReferencesFactory : IWorkspaceService
+    {
+        DefinitionsAndReferences CreateDefinitionsAndReferences(
+            Solution solution, IEnumerable<ReferencedSymbol> referencedSymbols);
+    }
+
+    [ExportWorkspaceService(typeof(IDefinitionsAndReferencesFactory)), Shared]
+    internal class DefaultDefinitionsAndReferencesFactory : IDefinitionsAndReferencesFactory
```

Comment:
The majority of code here came from LibraryManager_FindReferences.cs

The filtering, ordering and adding of third party items can all happen now prior to presentation.